packages/lib/src/vis/hooks.ts

```diff
@@ -53,21 +53,34 @@ export function useDomains(
 
 export function useCameraState<T>(
   factory: (camera: Camera) => T,
-  equalityFn?: (prev: T, next: T) => boolean
+  deps: unknown[],
+  equalityFn = (prev: T, next: T) => Object.is(prev, next)
 ): T {
   const camera = useThree((state) => state.camera);
+  const rerender = useRerender();
 
-  const factoryRef = useSyncedRef(factory);
-  const [state, setState] = useState(() => factoryRef.current(camera));
+  const stateRef = useRef<T>(); // ref instead of state to avoid re-render when deps change
+  const factoryRef = useSyncedRef(factory); // ensure `useMemo` always sees latest `factory` reference
+
+  useMemo(() => {
+    // Compute state synchronously when deps change
+    stateRef.current = factoryRef.current(camera);
+  }, deps); // eslint-disable-line react-hooks/exhaustive-deps
 
   useFrame(() => {
+    // Recompute state and re-render on every frame
     const next = factoryRef.current(camera);
-    if (!equalityFn || !equalityFn(state, next)) {
-      setState(next);
+
+    // ... unless state hasn't change
+    if (equalityFn(stateRef.current as T, next)) {
+      return;
     }
+
+    stateRef.current = next;
+    rerender();
```

Comment:
`useRerender` allows us to split the state update and the re-rendering, which is impossible to do with `useState`. This way we can update the state in `useMemo` synchronously, without re-rendering.